File: LeetcodeJS-problems/Arrays/8-group-anagrams.js
// Input: strs = ["act", "pots", "tops", "cat", "stop", "hat"]

// Output: [["hat"], ["act", "cat"], ["stop", "pots", "tops"]]


// Sorting
class Solution {
    groupAnagrams(strs) {
        const res = {};

        for (let s of strs) {
            const sortedS = s.split('').sort().join('')
            if (!res[sortedS]) {
                res[sortedS] = []
            }
            res[sortedS].push(s)
        }
        return Object.values(res);
    }
}

Input: strs = ["act", "pots", "tops", "cat", "stop", "hat"]



const solution = new Solution()
const app = solution.groupAnagrams(strs)


console.log(app)


// Hash Table

class Solution2 {

    groupAnagrams(strs) {
        const res = {}
        for (let s of strs) {
            const count = new Array(26).fill(0)
            for (let c of s) {
                count[c.charCodeAt(0) - 'a'.charCodeAt(0)] += 1
            }

            const key = count.join(',')
            if (!res[key]) {
                res[key] = []
            }
            res[key].push(s)
        }
        return Object.values(res);
    }
}


const solution2 = new Solution()
const apply2 = solution.groupAnagrams(strs)


console.log(apply2)

File: LeetcodeJS-problems/Arrays/8-group-anagrams.py


# Sorting
class Solution:
    def groupAnagrams(self, strs):
        res = {}
        for s in strs:
            sorted_s = ''.join(sorted(s))
            if sorted_s not in res:
                res[sorted_s] = []
            res[sorted_s].append(s)

        return list(res.values())


# Hash Table

class Solution2:
    def groupAnagrams(self, strs):
        res = defaultdict(list)
        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord('a')] += 1

            key = tuple(count)
            if key not in res:
                res[key] = []
            res[tuple(count)].append(s)

        return list(res.values())


# Prueba con el input

strs = ["act", "pots", "tops", "cat", "stop", "hat"]

solution = Solution()
test = solution.groupAnagrams(strs)
print(test, '<---- test1')

solution2 = Solution()
test2 = solution2.groupAnagrams(strs)
print(test2, '<---- test2')
